File: src/main/java/org/moinex/ui/common/CalendarController.java
/*
 * Filename: CalendarController.java
 * Created on: March  1, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.ui.common;

import java.time.LocalDate;
import java.time.Month;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import javafx.fxml.FXML;
import javafx.geometry.HPos;
import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.geometry.VPos;
import javafx.scene.control.Label;
import javafx.scene.layout.GridPane;
import javafx.scene.layout.HBox;
import javafx.scene.layout.Priority;
import javafx.scene.layout.Region;
import javafx.scene.layout.StackPane;
import javafx.scene.layout.VBox;
import javafx.scene.paint.Color;
import javafx.scene.shape.Circle;
import javafx.scene.text.Text;
import lombok.NoArgsConstructor;
import org.moinex.entities.CalendarEvent;
import org.moinex.services.CalendarService;
import org.moinex.ui.dialog.AddCalendarEventController;
import org.moinex.util.Constants;
import org.moinex.util.WindowUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.stereotype.Controller;

/**
 * Controller for the calendar view
 */
@Controller
@NoArgsConstructor
public class CalendarController
{
    @FXML
    private Label currentMonth;

    @FXML
    private GridPane calendar;

    private ConfigurableApplicationContext springContext;

    private LocalDate dateFocus;

    private LocalDate today;

    private List<CalendarEvent> calendarEvents;

    private CalendarService calendarService;

    /**
     * Constructor for CalendarController
     * @param calendarService The service for the calendar
     */
    @Autowired
    public CalendarController(CalendarService calendarService, ConfigurableApplicationContext springContext)
    {
        this.calendarService = calendarService;
        this.springContext = springContext;
    }

    @FXML
    public void initialize()
    {
        dateFocus = LocalDate.now();
        today     = LocalDate.now();

        loadCalendarEventsFromDatabase();

        drawCalendar();
    }

    @FXML
    private void handleBackOneMonth()
    {
        dateFocus = dateFocus.minusMonths(1);
        calendar.getChildren().clear();
        drawCalendar();
    }

    @FXML
    private void handleForwardOneMonth()
    {
        dateFocus = dateFocus.plusMonths(1);
        calendar.getChildren().clear();
        drawCalendar();
    }

    @FXML
    private void handleAddEvent()
    {
        WindowUtils.openModalWindow(Constants.ADD_CALENDAR_EVENT_FXML,
                                    "Add Calendar Event",
                                    springContext,
                                    (AddCalendarEventController controller)
                                        -> {},
                                    List.of(this::drawCalendar));
    }

    /**
     * Load the calendar events from the database
     */
    private void loadCalendarEventsFromDatabase()
    {
        calendarEvents = calendarService.getAllEvents();
    }

    /**
     * Draw the calendar grid
     */
    private void drawCalendar()
    {
        final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MMM/yy");
        currentMonth.setText(dateFocus.format(formatter));

        Integer monthMaxDate = dateFocus.getMonth().maxLength();

        if (!dateFocus.isLeapYear() && dateFocus.getMonth() == Month.FEBRUARY)
        {
            monthMaxDate = Constants.NON_LEAP_YEAR_FEBRUARY_DAYS;
        }

        Integer dateOffset =
            LocalDate.of(dateFocus.getYear(), dateFocus.getMonthValue(), 1)
                .getDayOfWeek()
                .getValue();

        // Adjust the date offset to start from Sunday at the first line of the calendar
        if (dateOffset.equals(Constants.WEEK_DAYS))
        {
            dateOffset = 0;
        }

        Integer currentDate    = 1;
        int totalGridCells = dateOffset + monthMaxDate;
        int totalRows =
            (int)Math.ceil(totalGridCells / Constants.WEEK_DAYS.doubleValue());

        Map<Integer, List<CalendarEvent>> calendarEventMap =
            getCalendarEventsMonth(dateFocus);

        calendar.getChildren().clear();

        Double calendarWidth  = calendar.getPrefWidth();
        double calendarHeight = calendar.getPrefHeight();

        // Create the weekday labels
        for (int j = 0; j < Constants.WEEKDAY_ABBREVIATIONS.length; j++)
        {
            Text dayLabel = new Text(Constants.WEEKDAY_ABBREVIATIONS[j]);
            dayLabel.setFont(Constants.CALENDAR_WEEKDAY_FONT_CONFIG);

            StackPane dayContainer = new StackPane(dayLabel);
            dayContainer.setPrefWidth(calendarWidth / Constants.WEEK_DAYS);
            dayContainer.setPrefHeight(calendarHeight / (totalRows + 1));
            GridPane.setHalignment(dayLabel, HPos.CENTER);
            GridPane.setValignment(dayLabel, VPos.CENTER);
            calendar.add(dayContainer, j, 0);
        }

        // Create the calendar cells for each day of the month
        for (int i = 0; i < totalRows; i++)
        {
            for (int j = 0; j < Constants.WEEK_DAYS; j++)
            {
                VBox cell = new VBox();
                cell.setMinSize(calendarWidth / Constants.WEEK_DAYS,
                                calendarHeight / (totalRows + 1));

                cell.setAlignment(Pos.TOP_CENTER);

                // Define the border style for the cell
                Double top    = (i == 0) ? Constants.CALENDAR_CELL_EXTERNAL_BORDER_WIDTH
                                         : Constants.CALENDAR_CELL_BORDER_WIDTH;
                Double bottom = (i == totalRows - 1)
                                    ? Constants.CALENDAR_CELL_EXTERNAL_BORDER_WIDTH
                                    : Constants.CALENDAR_CELL_BORDER_WIDTH;
                Double left   = (j == 0) ? Constants.CALENDAR_CELL_EXTERNAL_BORDER_WIDTH
                                         : Constants.CALENDAR_CELL_BORDER_WIDTH;
                Double right  = (j == Constants.WEEK_DAYS - 1)
                                    ? Constants.CALENDAR_CELL_EXTERNAL_BORDER_WIDTH
                                    : Constants.CALENDAR_CELL_BORDER_WIDTH;

                String borderStyle = String.format(
                    "-fx-border-color: black; "
                        + "-fx-border-width: %.1fpx %.1fpx %.1fpx %.1fpx;",
                    top,
                    right,
                    bottom,
                    left);

                cell.setStyle(borderStyle);

                if ((i == 0 && j >= dateOffset) ||
                    (i > 0 && currentDate <= monthMaxDate))
                {
                    Text dateText = new Text(String.valueOf(currentDate));
                    dateText.setFont(Constants.CALENDAR_DATE_FONT_CONFIG);

                    // Create the event indicators for the cell
                    HBox eventIndicators = new HBox(3);
                    eventIndicators.setAlignment(Pos.CENTER);

                    List<CalendarEvent> calendarEvents =
                        calendarEventMap.get(currentDate);
                    if (calendarEvents != null)
                    {
                        for (CalendarEvent event : calendarEvents)
                        {
                            Circle indicator = new Circle(
                                4,
                                Color.web(event.getEventType().getColorHex()));
                            eventIndicators.getChildren().add(indicator);
                        }
                    }

                    // Make the cell clickable to open the event details
                    // TODO: Implement the event details view
                    final Integer selectedDate = currentDate;
                    cell.setOnMouseClicked(
                        event -> System.out.println("Selected date: " + selectedDate));

                    // Align the event indicators to the bottom of the cell
                    Region spacer = new Region();
                    VBox.setVgrow(spacer, Priority.ALWAYS);
                    eventIndicators.setPadding(new Insets(0, 0, 5, 0));

                    cell.getChildren().addAll(dateText, spacer, eventIndicators);

                    // Highlight the current date
                    if (today.getYear() == dateFocus.getYear() &&
                        today.getMonth() == dateFocus.getMonth() &&
                        today.getDayOfMonth() == currentDate)
                    {
                        cell.setStyle("-fx-border-color: blue; -fx-border-width: "
                                      + "2px; -fx-background-color: lightblue;");
                    }

                    currentDate++;
                }

                calendar.add(cell, j, i + 1);
            }
        }
    }

    /**
     * Create a map of calendar events by date
     * @param calendarEvents The list of calendar events
     * @return A map of calendar events by date
     */
    private Map<Integer, List<CalendarEvent>>
    createCalendarMap(List<CalendarEvent> calendarEvents)
    {
        Map<Integer, List<CalendarEvent>> calendarEventMap = new HashMap<>();

        for (CalendarEvent event : calendarEvents)
        {
            Integer eventDate = event.getDate().getDayOfMonth();
            if (!calendarEventMap.containsKey(eventDate))
            {
                calendarEventMap.put(eventDate, List.of(event));
            }
            else
            {
                List<CalendarEvent> oldListByDate = calendarEventMap.get(eventDate);

                List<CalendarEvent> newList = new ArrayList<>(oldListByDate);
                newList.add(event);
                calendarEventMap.put(eventDate, newList);
            }
        }

        return calendarEventMap;
    }

    /**
     * Get the calendar events for the month
     * @param dateFocus The date to focus on
     * @return A map of calendar events by date
     */
    private Map<Integer, List<CalendarEvent>>
    getCalendarEventsMonth(LocalDate dateFocus)
    {
        List<CalendarEvent> calendarEventsMonth = new ArrayList<>();

        for (CalendarEvent event : calendarEvents)
        {
            if (event.getDate().getMonth() == dateFocus.getMonth() &&
                event.getDate().getYear() == dateFocus.getYear())
            {
                calendarEventsMonth.add(event);
            }
        }

        return createCalendarMap(calendarEventsMonth);
    }
}

File: src/main/kotlin/org/moinex/ui/common/CalendarController.kt
/*
 * Filename: CalendarController.kt (original filename: CalendarController.java)
 * Created on: March  1, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 19/03/2026
 */

package org.moinex.ui.common

import javafx.fxml.FXML
import javafx.geometry.HPos
import javafx.geometry.Insets
import javafx.geometry.Pos
import javafx.geometry.VPos
import javafx.scene.control.Label
import javafx.scene.layout.GridPane
import javafx.scene.layout.HBox
import javafx.scene.layout.Priority
import javafx.scene.layout.Region
import javafx.scene.layout.StackPane
import javafx.scene.layout.VBox
import javafx.scene.paint.Color
import javafx.scene.shape.Circle
import javafx.scene.text.Text
import org.moinex.common.constant.Constants
import org.moinex.common.constant.Files
import org.moinex.common.constant.TranslationKeys
import org.moinex.common.util.UIUtils
import org.moinex.common.util.WindowUtils
import org.moinex.model.CalendarEvent
import org.moinex.service.CalendarService
import org.moinex.service.PreferencesService
import org.moinex.ui.dialog.AddCalendarEventController
import org.springframework.context.ConfigurableApplicationContext
import org.springframework.stereotype.Controller
import java.time.LocalDate
import java.time.Month
import kotlin.math.ceil

@Controller
class CalendarController(
    private val calendarService: CalendarService,
    private val preferencesService: PreferencesService,
    private val springContext: ConfigurableApplicationContext,
) {
    @FXML
    private lateinit var currentMonth: Label

    @FXML
    private lateinit var calendar: GridPane

    private var dateFocus = LocalDate.now()
    private val today = LocalDate.now()
    private var calendarEvents = emptyList<CalendarEvent>()

    @FXML
    private fun initialize() {
        dateFocus = LocalDate.now()
        loadCalendarEventsFromDatabase()
        drawCalendar()
    }

    @FXML
    private fun handleBackOneMonth() {
        dateFocus = dateFocus.minusMonths(1)
        calendar.children.clear()
        drawCalendar()
    }

    @FXML
    private fun handleForwardOneMonth() {
        dateFocus = dateFocus.plusMonths(1)
        calendar.children.clear()
        drawCalendar()
    }

    @FXML
    private fun handleAddEvent() {
        WindowUtils.openModalWindow(
            Files.ADD_CALENDAR_EVENT_FXML,
            preferencesService.translate(TranslationKeys.COMMON_CALENDAR_MODAL_ADD_EVENT),
            springContext,
            { _: AddCalendarEventController -> },
            listOf(Runnable { drawCalendar() }),
        )
    }

    private fun loadCalendarEventsFromDatabase() {
        calendarEvents = calendarService.getAllEvents()
    }

    private fun drawCalendar() {
        val formatter = UIUtils.getShortMonthYearFormatter(preferencesService.locale)
        currentMonth.text = dateFocus.format(formatter)

        val monthMaxDate =
            if (!dateFocus.isLeapYear && dateFocus.month == Month.FEBRUARY) {
                Constants.NON_LEAP_YEAR_FEBRUARY_DAYS
            } else {
                dateFocus.month.maxLength()
            }

        var dateOffset =
            LocalDate
                .of(dateFocus.year, dateFocus.monthValue, 1)
                .dayOfWeek
                .value

        if (dateOffset == Constants.WEEK_DAYS) {
            dateOffset = 0
        }

        val totalGridCells = dateOffset + monthMaxDate
        val totalRows = ceil(totalGridCells.toDouble() / Constants.WEEK_DAYS).toInt()

        val calendarEventMap = getCalendarEventsMonth(dateFocus)

        calendar.children.clear()

        val calendarWidth = calendar.prefWidth
        val calendarHeight = calendar.prefHeight

        createWeekdayLabels(calendarWidth, calendarHeight, totalRows)
        createCalendarCells(
            calendarWidth,
            calendarHeight,
            totalRows,
            dateOffset,
            monthMaxDate,
            calendarEventMap,
        )
    }

    private fun createWeekdayLabels(
        calendarWidth: Double,
        calendarHeight: Double,
        totalRows: Int,
    ) {
        UIUtils.getWeekdayAbbreviations().forEachIndexed { j, weekday ->
            val dayLabel =
                Text(weekday).apply {
                    font = Constants.CALENDAR_WEEKDAY_FONT_CONFIG
                }

            val dayContainer =
                StackPane(dayLabel).apply {
                    prefWidth = calendarWidth / Constants.WEEK_DAYS
                    prefHeight = calendarHeight / (totalRows + 1)
                }

            GridPane.setHalignment(dayLabel, HPos.CENTER)
            GridPane.setValignment(dayLabel, VPos.CENTER)
            calendar.add(dayContainer, j, 0)
        }
    }

    private fun createCalendarCells(
        calendarWidth: Double,
        calendarHeight: Double,
        totalRows: Int,
        dateOffset: Int,
        monthMaxDate: Int,
        calendarEventMap: Map<Int, List<CalendarEvent>>,
    ) {
        var currentDate = 1

        repeat(totalRows) { i ->
            repeat(Constants.WEEK_DAYS) { j ->
                val cell = createCell(calendarWidth, calendarHeight, totalRows, i, j)

                if ((i == 0 && j >= dateOffset) || (i > 0 && currentDate <= monthMaxDate)) {
                    populateCell(cell, currentDate, calendarEventMap[currentDate])
                    currentDate++
                }

                calendar.add(cell, j, i + 1)
            }
        }
    }

    private fun createCell(
        calendarWidth: Double,
        calendarHeight: Double,
        totalRows: Int,
        row: Int,
        col: Int,
    ): VBox =
        VBox().apply {
            minWidth = calendarWidth / Constants.WEEK_DAYS
            minHeight = calendarHeight / (totalRows + 1)
            alignment = Pos.TOP_CENTER

            val borderWidths = calculateBorderWidths(row, col, totalRows)
            style =
                "-fx-border-color: black; " +
                "-fx-border-width: %.1fpx %.1fpx %.1fpx %.1fpx;".format(
                    borderWidths.top,
                    borderWidths.right,
                    borderWidths.bottom,
                    borderWidths.left,
                )
        }

    private fun calculateBorderWidths(
        row: Int,
        col: Int,
        totalRows: Int,
    ): BorderWidths =
        BorderWidths(
            top =
                if (row == 0) {
                    Constants.CALENDAR_CELL_EXTERNAL_BORDER_WIDTH
                } else {
                    Constants.CALENDAR_CELL_BORDER_WIDTH
                },
            bottom =
                if (row == totalRows - 1) {
                    Constants.CALENDAR_CELL_EXTERNAL_BORDER_WIDTH
                } else {
                    Constants.CALENDAR_CELL_BORDER_WIDTH
                },
            left =
                if (col == 0) {
                    Constants.CALENDAR_CELL_EXTERNAL_BORDER_WIDTH
                } else {
                    Constants.CALENDAR_CELL_BORDER_WIDTH
                },
            right =
                if (col == Constants.WEEK_DAYS - 1) {
                    Constants.CALENDAR_CELL_EXTERNAL_BORDER_WIDTH
                } else {
                    Constants.CALENDAR_CELL_BORDER_WIDTH
                },
        )

    private fun populateCell(
        cell: VBox,
        date: Int,
        events: List<CalendarEvent>?,
    ) {
        val dateText =
            Text(date.toString()).apply {
                font = Constants.CALENDAR_DATE_FONT_CONFIG
            }

        val eventIndicators =
            HBox(3.0).apply {
                alignment = Pos.CENTER
                events?.forEach { event ->
                    children.add(Circle(4.0, Color.web(event.eventType.colorHex)))
                }
            }

        val spacer =
            Region().apply {
                VBox.setVgrow(this, Priority.ALWAYS)
            }

        eventIndicators.padding = Insets(0.0, 0.0, 5.0, 0.0)
        cell.children.addAll(dateText, spacer, eventIndicators)

        if (isToday(date)) {
            cell.style = "-fx-border-color: blue; -fx-border-width: 2px; -fx-background-color: lightblue;"
        }
    }

    private fun isToday(date: Int): Boolean =
        today.year == dateFocus.year &&
            today.month == dateFocus.month &&
            today.dayOfMonth == date

    private fun createCalendarMap(calendarEvents: List<CalendarEvent>): Map<Int, List<CalendarEvent>> =
        calendarEvents.groupBy { it.date.dayOfMonth }

    private fun getCalendarEventsMonth(dateFocus: LocalDate): Map<Int, List<CalendarEvent>> {
        val calendarEventsMonth =
            calendarEvents.filter {
                it.date.month == dateFocus.month && it.date.year == dateFocus.year
            }

        return createCalendarMap(calendarEventsMonth)
    }

    private data class BorderWidths(
        val top: Double,
        val right: Double,
        val bottom: Double,
        val left: Double,
    )
}
